packages/unmock-core/src/interceptor/node-interceptor.ts

```diff
@@ -0,0 +1,106 @@
+import debug from "debug";
+import {
+  ClientRequest,
+  IncomingMessage,
+  RequestOptions,
+  ServerResponse,
+} from "http";
+import Mitm = require("mitm");
+import * as net from "net";
+import { handleRequest } from "../backend";
+import { ISerializedRequest, ISerializedResponse } from "../interfaces";
+import { serializeRequest } from "../serialize";
+import { IInterceptor, IInterceptorConstructor, IInterceptorOptions } from "./";
+import ClientRequestTracker from "./client-request-tracker";
+
+const debugLog = debug("unmock:node-interceptor");
+
+const respondFromSerializedResponse = (
+  serializedResponse: ISerializedResponse,
+  res: ServerResponse,
+) => {
+  res.writeHead(serializedResponse.statusCode, serializedResponse.headers);
+  res.end(serializedResponse.body);
+};
+
+interface IBypassableSocket extends net.Socket {
+  bypass: () => void;
+}
+
+/**
+ * Node.js interceptor using node-mitm.
+ * @param options Interceptor options
+ */
+const NodeInterceptorConstructor: IInterceptorConstructor = class NodeInterceptor
```

Comment:
Nitpicking, but why assign the class to a `const` instead of just declaring the class?
That is, if one still has to do `new NodeInterceptorConstructor(...)`, wouldn't it make more sense to have `new NodeInterceptor(...)`? I'd expect the former to be called once more to create a `NodeInterceptor`.

Reply:
Yep the idea is simply that because creating the interceptor starts the interception (without a separate "initialize" call), I wanted to add typing to the constructor itself with `IInterceptorConstructor`. And AFAIK one cannot implement an interface with class constructor directly so the constructor is typed here by assigning the class. Maybe it's overengineering but I don't see harm in it.

Reply:
I mean this does not work:
```
interface IInterceptor {
  new (options: IInterceptorOptions): IInterceptor;
  disable(): void;
}
```
but one needs `IInterceptorConstructor`. And with the `const` assignment it can be made explicit that the `NodeInterceptor` static class implements `IInterceptorConstructor`.

Reply:
That's fine, I was mostly referring to the naming convention.
Calling `new NodeInterceptorConstructor(...)` IMO implies we're creating a new constructor, where as here it creates a new `NodeInterceptor`.

As for the interfaces - I don't mind much; I'm unaware of languages that allow interfaces for the constructor. I guess the more typical solution would be to use abstract classes though.

Reply:
Ah ok, I'll rename it in the next PR.